transmart-core-db/src/main/groovy/org/transmartproject/db/accesscontrol/AccessControlChecks.groovy

```diff
@@ -222,99 +224,77 @@ class AccessControlChecks implements AuthorisationChecks, LegacyAuthorisationChe
+        criteria.add(Restrictions.in('secureObjectToken', accessibleStudyTokens))
+        criteria.getExecutableCriteria(sessionFactory.currentSession).list() as List<MDStudy>
     }
 
     /**
-     * Gets all tokens to the studies user have access to including PUBLIC tokens.
+     * Retrieves study ids for the studies to which the user has at least the required access level.
      * @param user
+     * @param requiredAccessLevel
+     * @return the set of study ids.
      */
-    private static Set<String> getAccessibleStudyTokensForUser(User user) {
-        user.studyToPatientDataAccessLevel.keySet() + PUBLIC_TOKENS
+    private static Set<String> getAccessibleStudyTokensForUser(User user, PatientDataAccessLevel requiredAccessLevel) {
+        user.studyToPatientDataAccessLevel.entrySet().stream()
+                .filter({ Map.Entry<String, PatientDataAccessLevel> entry ->
+                    entry.value >= requiredAccessLevel
+                })
+                .map({ Map.Entry<String, PatientDataAccessLevel> entry -> entry.key })
+                .collect(Collectors.toSet()) + PUBLIC_TOKENS
     }
 
-    private boolean exists(org.hibernate.criterion.DetachedCriteria criteria) {
+    private boolean exists(DetachedCriteria criteria) {
         (criteria.getExecutableCriteria(sessionFactory.currentSession).setMaxResults(1).uniqueResult() != null)
     }
 
-    /**
-     * Checks if the user has access to the (study-linked) dimensions in the list.
-     * @param dimensions
-     * @param user
-     * @throws AccessDeniedException if there are any dimensions in the list that the user does not have access to.
-     */
-    void checkDimensionsAccess(Collection<Dimension> dimensions, User user) throws AccessDeniedException {
-        def studies = getDimensionStudiesForUser(user)
-        def allowedDimensionNames = studies.collectMany { it.dimensions.collect { it.name }} as Set<String>
-        def dimensionNames = dimensions*.name as Set<String>
-        def deniedDimensions = dimensionNames - allowedDimensionNames
-        if (!deniedDimensions.empty) {
-            throw new AccessDeniedException("Access denied to dimensions: ${deniedDimensions.join(', ')}")
+    boolean canAccessConceptByCode(User user, PatientDataAccessLevel requiredAccessLevel, String conceptCode) {
+        if (conceptCode == null && conceptCode.empty) {
+            throw new AccessDeniedException('No concept code provided.')
         }
+
+        if (user.admin) {
+            return true
+        }
+
+        Set<String> tokens = getAccessibleStudyTokensForUser(user, requiredAccessLevel)
+
+        def conceptCriteria = DetachedCriteria.forClass(ConceptDimension)
+        conceptCriteria = conceptCriteria.add(StringUtils.like('conceptCode', conceptCode, MatchMode.EXACT))
+        conceptCriteria = conceptCriteria.setProjection(Projections.property('conceptPath'))
+
+        def criteria = DetachedCriteria.forClass(I2b2Secure)
```

Comment:
There is a possibility for reuse. Split this method on two: 1. getting `I2b2Secure` node by the concept code (weird part) 2. Check whether the user has access to the node (reuse `boolean hasAccess(User user, OntologyTerm term)`)

Reply:
Reuse, but also doing two queries instead of one

Reply:
There can be many nodes for the same concept.